asserts/preseed.go

```diff
@@ -0,0 +1,231 @@
+	return p.HeaderString("model")
+}
+
+// SystemLabel returns the label of the seeded system.
+func (p *Preseed) SystemLabel() string {
+	return p.HeaderString("system-label")
+}
+
+// Timestamp returns the time when the preseed assertion was issued.
+func (p *Preseed) Timestamp() time.Time {
+	return p.timestamp
+}
+
+// PreseedSHA3_384 returns the checksum of preseeding artifact.
+func (p *Preseed) PreseedSHA3_384() string {
+	return p.HeaderString("preseed-sha3-384")
+}
+
+// Snaps returns the snaps for preseeding.
+func (p *Preseed) Snaps() []*PreseedSnap {
+	return p.snaps
+}
+
+func checkPreseedSnap(snap map[string]interface{}) (*PreseedSnap, error) {
+	name, err := checkNotEmptyStringWhat(snap, "name", "of snap")
+	if err != nil {
+		return nil, err
+	}
+	if err := naming.ValidateSnap(name); err != nil {
+		return nil, fmt.Errorf("invalid snap name %q", name)
+	}
+
+	what := fmt.Sprintf("of snap %q", name)
+
+	// snap id can be omitted if the model allows for unasserted snaps
+	var snapID string
+	if _, ok := snap["id"]; ok {
+		snapID, err = checkStringMatchesWhat(snap, "id", what, naming.ValidSnapID)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var snapRevision int
+	if _, ok := snap["revision"]; ok {
+		var err error
+		snapRevision, err = checkSnapRevisionWhat(snap, "revision", what)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	if snapID != "" && snapRevision <= 0 {
+		return nil, fmt.Errorf("snap revision is required when snap id is set")
+	}
+	if snapID == "" && snapRevision > 0 {
+		return nil, fmt.Errorf("snap id is required when revision is set")
+	}
+
+	return &PreseedSnap{
+		Name:     name,
+		SnapID:   snapID,
+		Revision: snapRevision,
+	}, nil
+}
+
+func checkPreseedSnaps(snapList interface{}) ([]*PreseedSnap, error) {
+	const wrongHeaderType = `"snaps" header must be a list of maps`
+
+	entries, ok := snapList.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf(wrongHeaderType)
+	}
+
+	seen := make(map[string]bool, len(entries))
+	seenIDs := make(map[string]string, len(entries))
+	snaps := make([]*PreseedSnap, 0, len(entries))
+	for _, entry := range entries {
+		snap, ok := entry.(map[string]interface{})
+		if !ok {
+			return nil, fmt.Errorf(wrongHeaderType)
+		}
+		preseedSnap, err := checkPreseedSnap(snap)
+		if err != nil {
+			return nil, err
+		}
+
+		if seen[preseedSnap.Name] {
+			return nil, fmt.Errorf("cannot list the same snap %q multiple times", preseedSnap.Name)
+		}
+		seen[preseedSnap.Name] = true
+		snapID := preseedSnap.SnapID
+		if underName := seenIDs[snapID]; underName != "" {
+			return nil, fmt.Errorf("cannot specify the same snap id %q multiple times, specified for snaps %q and %q", snapID, underName, preseedSnap.Name)
+		}
+		seenIDs[snapID] = preseedSnap.Name
+		snaps = append(snaps, preseedSnap)
+	}
+
+	return snaps, nil
+}
+
+func assemblePreseed(assert assertionBase) (Assertion, error) {
+	_, err := checkStringMatches(assert.headers, "brand-id", validAccountID)
```

Comment:
this check doesn't hurt but is not strictly needed because brand-id must match authority-id and authority-id must be used to point to a valid account-key that needs a matching account assertion, that chaining assures that brand-id is a valid accoun-id, model doesn't have this check

Reply:
Hmm, maybe I'm confused, but a test like this passes without such check:

```
func (ps *preseedSuite) TestSnapXXX(c *C) {
	encoded := strings.Replace(preseedExample, "TSLINE", ps.tsLine, 1)
	encoded = strings.Replace(encoded, "OTHER", "", 1)
	encoded = strings.Replace(encoded, "authority-id: brand-id1", "authority-id: _#$", 1)
	encoded = strings.Replace(encoded, "brand-id: brand-id1", "brand-id: _#$", 1)

	a, err := asserts.Decode([]byte(encoded))
	c.Assert(err, IsNil)

	c.Check(a.Type(), Equals, asserts.PreseedType)
	preseed := a.(*asserts.Preseed)
	c.Check(preseed.AuthorityID(), Equals, "_#$")
	c.Check(preseed.BrandID(), Equals, "_#$")
}
```

Reply:
Got it, I'm fine to drop it then to be consistent. Thanks

Reply:
It's dropped now.